engine/robot_logic/traversal.py

```diff
@@ -208,64 +212,98 @@ def update_state(robot_state, velocity, omega):
-            [curr_x, curr_y, curr_head],
-            [robot_state.move_dist * robot_state.time_step, 0],
-        )
+        curr_pos = robot_state.state
+        vel = calculate_dist(past_pos[:2], curr_pos[:2]) / (dt / 1000)
+        if robot_state.using_ekf:
+            [new_x, new_y, _] = robot_state.ekf.get_predicted_state(
+                curr_pos, [vel * dt / 1000, 0]
+            )
+        else:
+            [new_x, new_y, _] = curr_pos + vel * dt / 1000
         next_radius = calculate_dist(base_station_loc, (new_x, new_y))
         robot_state.goal_location = (new_x, new_y)
         # if moving will cause the robot to move through the obstacle
-        is_next_timestep_blocked = next_radius < robot_state.detect_obstacle_range
+        if not robot_state.is_sim:
+            # if obstacle detected; have to make sure move_dist < detect_obstacle_range; otherwise, robot might hit obstacle before we go back to this loop
+            is_next_timestep_blocked = (
+                front_ultrasonic.distance() < robot_state.detect_obstacle_range
+            )
+        else:
+            next_radius_2 = calculate_dist(base_station_loc, curr_pos[:2])
+            threshold = 1
+            is_next_timestep_blocked = abs(next_radius_2 - roomba_radius) < threshold
         # sensor should not detect something in the robot
         if (next_radius > roomba_radius) or is_next_timestep_blocked:
             # this needs to be synchronous/PID'ed, otherwise, turn might be called while robot moving forward
-            if robot_state.is_sim:
-                # for some reason I don't think this should work. This needs to be blocking: wait for the robot to finish going backward before turning
-                travel(robot_state, -robot_state.move_dist, 0)
-                travel(robot_state, 0, robot_state.turn_angle)
-            else:
-                robot_state.motor_controller.motors(0, 0)
-                # TODO: change this to pid or time based. NEED TO MAKE SURE ROBOT DOESN'T BREAK WHEN GOING FROM
-                #  POS VEL TO NEG VEL IN A SHORT PERIOD OF TIME -> ramp down prob
+            allowed_dist_error = 0.01
+            allowed_heading_error = 0.01
+            # double check that robot wont collide behind
+            move_to_target_node(
+                robot_state,
+                (
+                    curr_pos[0] - robot_state.move_dist * math.cos(curr_pos[2]),
+                    curr_pos[1] - robot_state.move_dist * math.sin(curr_pos[2]),
+                ),
+                allowed_dist_error,
+                database,
+            )
+            turn_to_target_heading(
+                robot_state,
+                curr_pos[2] + robot_state.turn_angle,
+                allowed_heading_error,
+                database,
+            )
         else:
-            if robot_state.is_sim:
-                travel(robot_state, robot_state.move_dist, 0)
-            else:
+            travel(robot_state, robot_state.move_dist, 0)
```

Comment:
in this else clause aren't we supposed to move toward the goal location (robot_state.goal_location = (new_x, new_y))? not sure if we should use move_to_target_node instead or if the changes you're making to travel will take this into account?

Reply:
we aren't moving to the goal location, we just want to move forward. Using move_to_target_node doesn't really make sense because we're just moving forward; there is no end goal. We could set a target to be a few "insert length" in front of us, but using PID on every single instance of that length is slow (robot accelerates, then decelerates, etc)

Reply:
so with Roomba mode (and also with the travel method) we are disregarding the idea of moving from waypoint to waypoint? Then we should prolly remove this line: robot_state.goal_location = (new_x, new_y).  Also why do we use new_x and new_y to check whether or not the robot would crash into an obstacle/is within the Roomba radius if the robot isn't even moving to this new position?

Reply:
yeah, i could remove goal_location, but do we want to not update it? if we dont, it no longer "keep its invariant" or whatever. new_x and new_y is where I predict the robot to be, not necessarily where it will go, and also not where it should be (EKF). Otherwise, if we cant predict where the robot will be, we cant determine if it will go out of bounds until its already out of bounds